### PR TITLE
feat(admin-ui + admin-api): connect apps api with ui

### PR DESCRIPTION
**Pull request by frdomovic**
# feat(admin-ui + admin-api): connect apps api with ui
 
Screenshot 2024-04-15 at 18 17 32

### CODE REVIEW

File: node-ui/src/components/applications/ApplicationsTable.jsx [hunk 1]
```diff
@@ -129,16 +129,18 @@ export function ApplicationsTable({ applications, install, uninstall }) {
...
-                  <MenuIconDropdown
-                    options={[
-                      {
-                        buttonText: t.uninstallButtonText,
-                        onClick: () => uninstall(id),
-                      },
-                    ]}
-                  />
-                </div>
+                {false && (
```
Comment (frdomovic) [inline node-ui/src/components/applications/ApplicationsTable.jsx:132]: not implemented

File: crates/server/src/admin/service.rs [hunk 2]
```diff
@@ -60,14 +61,17 @@ pub(crate) fn setup(
...
         application_dir: config.application_dir.clone(),
     });
+
     let admin_router = Router::new()
         .route("/health", get(health_check_handler))
         .route("/root-key", post(create_root_key_handler))
         .route("/request-challenge", post(request_challenge_handler))
         .route("/install-application", post(install_application_handler))
-        .layer(Extension(state))
+        .layer(Extension(state.clone()))
```
Comment (MatejVukosav) [inline crates/server/src/admin/service.rs]: I have updated this part in the new PR so we have shared state https://github.com/calimero-network/core/pull/154

File: crates/server/src/admin/service.rs [hunk 3]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
...
+    let application_base_path = dir.join(application_id.to_string());
+
+    if let Ok(entries) = fs::read_dir(&application_base_path) {
+        let mut versions_with_binary = entries
+            .filter_map(|entry| {
+                let entry = entry.ok()?;
+                let entry_path = entry.path();
+
+                let version = {
+                    semver::Version::parse(entry_path.file_name()?.to_string_lossy().as_ref())
```
Comment (MatejVukosav) [inline crates/server/src/admin/service.rs:430]: why `to_string_lossy()`?

File: crates/server/src/admin/service.rs [hunk 4]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
...
+                };
+
+                let binary_path = entry_path.join("binary.wasm");
+                binary_path.exists().then_some((version, entry_path))
+            })
+            .collect::<Vec<_>>();
+
+        versions_with_binary.sort_by(|a, b| b.0.cmp(&a.0));
+
+        if let Some((_, path)) = versions_with_binary.first() {
```
Comment (MatejVukosav) [inline crates/server/src/admin/service.rs]: First already returns none if there are no element
Check this
```
versions_with_binary.first()
            .map(|(_, path)| path.to_string_lossy().into_owned())
```

File: crates/server/src/admin/service.rs [hunk 5]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
...
+                    get_latest_application_path(&state.application_dir, &file_name);
+                if let Some(latest_version) = latest_version {
+                    let app_name = file_name.to_string();
+                    if let Some(version) = latest_version.rsplit('/').next() {
+                        applications.insert(app_name, version.to_string());
+                    }
+                }
+            }
+        });
+        let response_body = json!({
```
Comment (MatejVukosav) [inline crates/server/src/admin/service.rs]: Create struct and return it

File: node-ui/src/pages/Applications.jsx [hunk 6]
```diff
@@ -27,6 +29,21 @@ export default function Applications() {
     }
   }, [packages]);
 
+  useEffect(() => {
+    const setApps = async () => {
+      const applicationIds = Object.keys(
+        JSON.parse((await axios.get("/admin-api/applications")).data).apps
```
Comment (MatejVukosav) [inline node-ui/src/pages/Applications.jsx]: Encapsulate api call outside page. You can check only-peers frontend for reference

File: crates/server/src/admin/service.rs (hunk 2 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: Will rebease after merged

File: crates/server/src/admin/service.rs (hunk 3 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs:430]: to_string()  / to_str() / just as_ref() -> the trait bound `OsStr: AsRef<str>` is not satisfied

File: crates/server/src/admin/service.rs (hunk 4 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: Yep this works, updated

File: node-ui/src/pages/Applications.jsx (hunk 6 above)
Comment (frdomovic) [reply node-ui/src/pages/Applications.jsx]: Updated

File: crates/server/src/admin/service.rs (hunk 2 above)
Comment (miraclx) [inline crates/server/src/admin/service.rs]: why duplicate these entries?

Just put the extension at the end, all the handlers will be able to access it

File: crates/server/src/admin/service.rs (hunk 3 above)
Comment (miraclx) [reply crates/server/src/admin/service.rs:430]: OS file names are not guaranteed to be UTF-8 strings, so you have to attempt at either a lossy or fallible conversion

File: crates/server/src/admin/service.rs [hunk 7]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
     .into_response())
 }
 
+fn get_latest_application_path(dir: &camino::Utf8Path, application_id: &str) -> Option<String> {
+    let application_base_path = dir.join(application_id.to_string());
+
+    if let Ok(entries) = fs::read_dir(&application_base_path) {
```
Comment (miraclx) [inline crates/server/src/admin/service.rs:423]: sucks that we're doing blocking IO in an async environment, we can refactor after token

File: crates/server/src/admin/service.rs [hunk 8]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
     .into_response())
 }
 
+fn get_latest_application_path(dir: &camino::Utf8Path, application_id: &str) -> Option<String> {
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: Return an `Option<PathBuf>` here

File: crates/server/src/admin/service.rs [hunk 9]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
...
+                let binary_path = entry_path.join("binary.wasm");
+                binary_path.exists().then_some((version, entry_path))
+            })
+            .collect::<Vec<_>>();
+
+        versions_with_binary.sort_by(|a, b| b.0.cmp(&a.0));
+
+        versions_with_binary
+            .first()
+            .map(|(_, path)| path.to_string_lossy().into_owned())
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: not a healthy conversion, if the path is useless to read from the file

and ideally, we should be using [camino](https://docs.rs/camino) which will do a falible conversion to UTF-8 to get that guarantee, and we already have it in the codebase

And this already has serde impls (behind a feature flag), so you won't need to convert to `String` at all

File: crates/server/src/admin/service.rs [hunk 10]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
...
+
+async fn fetch_application_handler(
+    Extension(state): Extension<Arc<ServiceState>>,
+    session: Session,
+) -> impl IntoResponse {
+    if let Ok(entries) = fs::read_dir(&state.application_dir) {
+        let mut applications: HashMap<String, String> = HashMap::new();
+
+        entries.filter_map(|entry| entry.ok()).for_each(|entry| {
+            if let Some(file_name) = entry.file_name().to_str() {
```
Comment (miraclx) [inline crates/server/src/admin/service.rs:468]: maybe we add a log to the CLI that we're unable to read the application ID, so it doesn't silently ignore it

a lossy conversion to aid display will be necessary since we can only print utf-8 data

File: crates/server/src/admin/service.rs [hunk 11]
```diff
@@ -332,6 +336,72 @@ async fn install_application_handler(
...
+                let latest_version =
+                    get_latest_application_path(&state.application_dir, &file_name);
+                if let Some(latest_version) = latest_version {
+                    let app_name = file_name.to_string();
+                    if let Some(version) = latest_version.rsplit('/').next() {
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: looks like this is guatanteed to be true, unless you intend to use `get_latest_application_path` elsewhere in this admin page, you can just return the version itself.

File: crates/server/Cargo.toml [hunk 12]
```diff
@@ -25,6 +25,7 @@ rand.workspace = true
 reqwest.workspace = true
 serde = { workspace = true, features = ["derive"] }
 serde_json.workspace = true
+semver.workspace = true
```
Comment (miraclx) [inline crates/server/Cargo.toml]: nit: order

File: crates/server/src/admin/service.rs (hunk 2 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: Added at the end

File: crates/server/src/admin/service.rs (hunk 9 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: upated

File: crates/server/src/admin/service.rs (hunk 11 above)
Comment (frdomovic) [reply crates/server/src/admin/service.rs]: updated

File: crates/server/src/admin/service.rs [hunk 13]
```diff
@@ -336,7 +336,7 @@ async fn install_application_handler(
     .into_response())
 }
 
-fn get_latest_application_path(dir: &camino::Utf8Path, application_id: &str) -> Option<String> {
+fn get_latest_application_path(dir: &camino::Utf8Path, application_id: &str) -> Option<PathBuf> {
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: just saw the input here is `camino::Utf8Path`, please return `camino::Utf8PathBuf`

File: crates/server/src/admin/service.rs [hunk 14]
```diff
@@ -383,10 +386,15 @@ async fn fetch_application_handler(
                     get_latest_application_path(&state.application_dir, &file_name);
                 if let Some(latest_version) = latest_version {
                     let app_name = file_name.to_string();
-                    if let Some(version) = latest_version.rsplit('/').next() {
-                        applications.insert(app_name, version.to_string());
-                    }
+                    let version = latest_version
+                        .file_name()
+                        .unwrap()
+                        .to_str()
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: why this change?

Comment (miraclx) [reply crates/server/src/admin/service.rs]: to clarify, I meant return the version itself from the `get_latest_application_path` instead of a path since it's not used elsewhere

make it `find_latest_version`, and return the `(version, path)` pair you already have

File: crates/server/src/admin/service.rs [hunk 15]
```diff
@@ -332,6 +336,80 @@ async fn install_application_handler(
...
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        versions_with_binary.sort_by(|a, b| b.0.cmp(&a.0));
+
+        versions_with_binary
+            .first()
+            .map(|(_, path)| PathBuf::from(path))
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: that is, remove this

```suggestion
```
